src/macaron/slsa_analyzer/package_registry/npm_registry.py

```diff
@@ -0,0 +1,265 @@
+            res_obj = response.json()
+        except requests.exceptions.JSONDecodeError as error:
+            raise InvalidHTTPResponseError(f"Failed to process response from Maven central for {url}.") from error
+        if not res_obj:
+            raise InvalidHTTPResponseError(f"Empty response returned by {url} .")
+        if not res_obj.get("attestations"):
+            raise InvalidHTTPResponseError(f"The response returned by {url} misses `attestations` attribute.")
+
+        # Download the SLSA provenance only.
+        for att in res_obj.get("attestations"):
+            if not att.get("predicateType"):
+                logger.debug("predicateType attribute is missing for %s", url)
+                continue
+            if att.get("predicateType") != "https://slsa.dev/provenance/v0.2":
+                logger.debug("predicateType %s is not accepted. Skipping...", att.get("predicateType"))
+                continue
+            if not (bundle := att.get("bundle")):
+                logger.debug("bundle attribute in the attestation is missing. Skipping...")
+                continue
+            if not (dsse_env := bundle.get("dsseEnvelope")):
+                logger.debug("dsseEnvelope attribute in the bundle is missing. Skipping...")
+                continue
+
+            try:
+                with open(download_path, "w", encoding="utf-8") as file:
+                    json.dump(dsse_env, file)
+                    return True
+            except OSError as error:
+                logger.debug(
+                    "Failed to write the downloaded attestation from %s to %s. Error: %s",
+                    url,
+                    download_path,
+                    error,
+                )
+
+        return False
+
+
+class NPMAttestationAsset(NamedTuple):
+    """An attestation asset hosted on the npm registry.
+
+    The API Documentation can be found here:
+    """
+
+    #: The namespace of the artifact on npm.
+    namespace: str | None
+
+    #: The artifact ID.
+    artifact_id: str
+
+    #: The version of the asset.
+    version: str
+
+    #: The npm registry.
+    npm_registry: NPMRegistry
+
+    #: The size of the asset (in bytes). This attribute is added to match the AssetLocator
+    #: protocol and is not used because npm API registry does not provide it.
+    size_in_bytes: int
+
+    @property
+    def name(self) -> str:
+        """Get the asset name."""
+        return self.artifact_id
+
+    @property
+    def url(self) -> str:
+        """Get the download URL of the asset.
+
+        Note: we assume that the path parameters used to construct the URL are sanitized already.
+
+        Returns
+        -------
+        str
+        """
+        # Build the path parameters.
+        path_params = [self.npm_registry.attestation_endpoint]
+        if self.namespace:
+            path_params.append(self.namespace)
+        path_params.append(self.artifact_id)
+        path = f'{"/".join(path_params)}'
```

Comment:
I think we can remove the f string here as `"/".join(path_params)` will return a string.